File: symmetric-tree.java
class GfG {
    // return true/false denoting whether the tree is Symmetric or not
    public static boolean isSymmetric(Node root) {
        if (root == null)
            return true; // Empty tree is symmetric

        return isSymmetricUtil(root.left, root.right);
    }

    private static boolean isSymmetricUtil(Node left, Node right) {
        // If both subtrees are empty, they are symmetric
        if (left == null && right == null)
            return true;

        // If one subtree is empty and the other is not, they are not symmetric
        if (left == null || right == null)
            return false;

        // Check if the values of the current nodes are equal
        // and if the left subtree's left child is equal to the right subtree's right child
        // and the left subtree's right child is equal to the right subtree's left child
        return (left.data == right.data) && isSymmetricUtil(left.left, right.right)
                && isSymmetricUtil(left.right, right.left);
    }
}

File: Difficulty: Easy/Symmetric Tree/symmetric-tree.cpp
/*
class Node {
public:
    int data;
    Node *left, *right;

    Node(int val) {
        data = val;
        left = right = nullptr;
    }
};
*/

class Solution {
  public:
    // Helper function to check if two trees are mirrors
    bool isMirror(Node* t1, Node* t2) {
        if (!t1 && !t2) return true;             // Both are null
        if (!t1 || !t2) return false;            // One is null, the other isn't
        if (t1->data != t2->data) return false;  // Data doesn't match

        // Check the mirror condition recursively
        return isMirror(t1->left, t2->right) && isMirror(t1->right, t2->left);
    }

    bool isSymmetric(Node* root) {
        if (!root) return true; // An empty tree is symmetric
        return isMirror(root->left, root->right);
    }
};
